diem-move/parallel-executor/src/scheduler.rs

```diff
@@ -1,113 +1,470 @@
 // Copyright (c) The Diem Core Contributors
 // SPDX-License-Identifier: Apache-2.0
-
-use crossbeam_queue::SegQueue;
-use mvhashmap::Version;
-use std::sync::{
-    atomic::{AtomicUsize, Ordering},
-    Arc, RwLock,
+use crossbeam::utils::CachePadded;
+use std::{
+    cmp::min,
+    hint,
+    sync::{
+        atomic::{AtomicBool, AtomicUsize, Ordering},
+        Mutex,
+    },
 };
 
-#[repr(usize)]
-enum ExecutionStatus {
-    Executed = 1,
-    NotExecuted = 0,
+// Type aliases.
+pub type TxnIndex = usize;
+pub type Incarnation = usize;
+pub type Version = (TxnIndex, Incarnation);
+
+// A struct to track the number of active tasks in the scheduler using RAII.
+pub struct TaskGuard<'a> {
+    counter: &'a AtomicUsize,
+}
+
+impl<'a> TaskGuard<'a> {
+    pub fn new(counter: &'a AtomicUsize) -> Self {
+        counter.fetch_add(1, Ordering::SeqCst);
+        Self { counter }
+    }
+}
+
+impl Drop for TaskGuard<'_> {
+    fn drop(&mut self) {
+        assert!(self.counter.fetch_sub(1, Ordering::SeqCst) > 0);
+    }
+}
+
+// A holder for potential task returned from the Scheduler. ExecutionTask and ValidationTask
+// each contain a version of transaction that must be executed or validated, respectively.
+// NoTask holds no task (similar None if we wrapped tasks in Option), and Done implies that
+// there are no more tasks and the scheduler is done.
+pub enum SchedulerTask<'a> {
+    ExecutionTask(Version, TaskGuard<'a>),
+    ValidationTask(Version, TaskGuard<'a>),
+    NoTask,
+    Done,
+}
+
+// All possible statuses for each transaction. Each status contains the latest incarnation number.
+//
+// 'ReadyToExecute' means that the corresponding incarnation should be executed and the scheduler
+// must eventually create a corresponding execution task. The scheduler ensures that exactly one
+// execution task gets created, changing the status to 'Executing' in the process.
+//
+// 'Executing' status of an incarnation turns into 'Executed' if the execution task finishes, or
+// if a dependency is encountered, it becomes 'ReadyToExecute(incarnation + 1)' once the
+// dependency is resolved. An 'Executed' status allows creation of validation tasks for the
+// corresponding incarnation, and a validation failure leads to an abort. The scheduler ensures
+// that there is exactly one abort, changing the status to 'Aborting' in the process. Once the
+// thread that successfully aborted performs everything that's required, it sets the status
+// to 'ReadyToExecute(incarnation + 1)', allowing the scheduler to create an execution
+// task for the next incarnation of the transaction.
+//
+// Status transition diagram when with no dependencies:
```

Comment:
I'd like to have some fancier diagram illustrating the state transitions.
```
Ready(i)
    | incarnate
    ↓                                          resume
Executing(i) (pending for exact one execution)   ------------> Ready(i+1)
    |  finish_execution
    ↓ 
Executed(i) (pending for (re)validations)
    | abort
    ↓    finish_validation (abort successfully)
Aborting    ---------------> Ready(i+1)
```

Reply:
nice!